File: 2021/java/665.non-decreasing-array.java
class Solution {
    public boolean checkPossibility(int[] nums) {
        if (nums == null || nums.length < 2) return true;
        int count = 0;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i - 1] > nums[i]) {
                count++;
                // reset nums[i-1] = nums[i], if current is decreasing, so that some cases, we can cover. e.g [3,4,2,3]
                if (i < 2 || nums[i - 2] <= nums[i]) nums[i - 1] = nums[i];
                else nums[i] = nums[i - 1];
            }
            
            if (count >= 2) return false;
        }
        
        return true;
    }
}

File: 2021/python/665.non-decreasing-array.py
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        if not nums or len(nums) < 2:
            return True
        count = 0
        for i in range(1, len(nums)):
            if nums[i - 1] > nums[i]:
                count += 1
                if i < 2 or nums[i - 2] <= nums[i]:
                    nums[i - 1] = nums[i]
                else:
                    nums[i] = nums[i - 1]
            if count > 1:
                return False
        return True
                
            
